drivers/wifi/esp8266/esp8266.h

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2019 Tobias Svehagen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_WIFI_ESP8266_H_
+#define ZEPHYR_INCLUDE_DRIVERS_WIFI_ESP8266_H_
+
+#include <kernel.h>
+#include <net/net_context.h>
+#include <net/net_if.h>
+#include <net/net_ip.h>
+#include <net/net_pkt.h>
+#include <net/wifi_mgmt.h>
+
+#include "modem_context.h"
+#include "modem_cmd_handler.h"
+#include "modem_iface_uart.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ESP8266_MAX_SOCKETS 5
+#define ESP8266_PKT_BUF_LEN 256
+
+/* Maximum amount that can be sent with CIPSEND and read with CIPRECVDATA */
+#define ESP8266_MTU		2048
+#define CIPRECVDATA_MAX_LEN	ESP8266_MTU
+
+#define INVALID_LINK_ID		255
+
+#define MDM_RING_BUF_SIZE	1024
+#define MDM_RECV_MAX_BUF	30
+#define MDM_RECV_BUF_SIZE	128
+#define CMD_BUF_ALLOC_TIMEOUT	K_SECONDS(1)
+
+#define ESP_CMD_TIMEOUT		K_SECONDS(10)
+#define ESP_SCAN_TIMEOUT	K_SECONDS(10)
+#define ESP_INIT_TIMEOUT	K_SECONDS(10)
+
+extern struct esp8266_data esp8266_driver_data;
+
+enum esp8266_socket_flags {
+	ESP_SOCK_VALID      = BIT(1),
+	ESP_SOCK_CONNECTING = BIT(2),
+	ESP_SOCK_CONNECTED  = BIT(3)
+};
+
+struct esp8266_socket {
+	/* internal */
+	u8_t idx;
+	u8_t link_id;
+	u8_t flags;
+
+	/* socket info */
+	sa_family_t family;
+	enum net_sock_type type;
+	enum net_ip_protocol ip_proto;
+	struct sockaddr src;
+	struct sockaddr dst;
+
+	/* for +CIPRECVDATA */
+	size_t bytes_avail;
+
+	/* packets */
+	struct k_fifo fifo_rx_pkt;
+	struct net_pkt *tx_pkt;
+
+	/* sem */
+	struct k_sem sem_data_ready;
+
+	/* work */
+	struct k_work connect_work;
+	struct k_work send_work;
+	struct k_work recv_work;
+	struct k_work recvdata_work;
+
+	/* net context */
+	struct net_context *context;
+	net_context_connect_cb_t connect_cb;
+	net_context_send_cb_t send_cb;
+	net_context_recv_cb_t recv_cb;
+
+	/* callback data */
+	void *conn_user_data;
+	void *send_user_data;
+	void *recv_user_data;
+};
+
+enum esp8266_data_flag {
+	EDF_STA_CONNECTING = BIT(1),
+	EDF_STA_CONNECTED  = BIT(2)
+};
+
+/* driver data */
+struct esp8266_data {
+	struct net_if *net_iface;
+
+	u8_t flags;
+
+	/* used to linearize net_pkt's during tx/rx */
+	u8_t pkt_buf[ESP8266_PKT_BUF_LEN];
+
+	/* addresses  */
+	struct in_addr ip;
```

Comment:
So IPv6 is not supported by the device?

Reply:
I believe both the esp8266 and esp32 support IPv6 but as a first release driver IPv4 is a good start.

Reply:
I have noticed that usually nobody is interested in adding IPv6 support later in various drivers / protocol libraries if that is not done from day one.

Reply:
I'll lend my support to add IPv6 functionality but I don't have access at present to a IPv6 router/AP.

Reply:
I do not believe IPv6 is supported in the AT firmware at least. I have not seen any commands that suggest IPv6. @WilliamGFish do you have any references for IPv6 support?

Reply:
Great question but as I have never used IPv6 apart from a singular POC years ago, no but I'll have a look.

Reply:
I have looked at the forums and the they are capable of IPv6 but only via LwIP in the ESP-IDF and ESP-SDK. There does not appear to be any mention of AT commands.

Poeple have said:"Unfortunately the documentation (mostly in Chinese) is far from complete over there and the documentation of the IPv6 example is virtually non-existent and appears to be centered on providing some form of IPv6 minimal capability for an ESP32 acting as a WiFi Station and AP simultaneously."